panther_docking/src/panther_charging_dock.cpp

```
@@ -0,0 +1,420 @@
+}
+
+bool PantherChargingDock::isCharging()
+{
+  std::shared_ptr<panther_msgs::msg::ChargingStatus> charging_status_msg;
+
+  if (!panther_utils::common_utilities::IsPantherVersionAtLeast(panther_version_, 1.2f)) {
+    try {
+      return isDocked();
+    } catch (const opennav_docking_core::FailedToDetectDock & e) {
+      return false;
+    }
+  }
+
+  // TODO: This might be used when a robot is undocking but we do not want to enable charging when
+  // the robot is undocking
+  try {
+    setChargerState(true);
+  } catch (const std::runtime_error & e) {
+    throw opennav_docking_core::FailedToCharge(
+      "An exception occurred while enabling charging: " + std::string(e.what()));
+  }
+
+  // CAUTION: The  controller frequency can be higher than the message frequency
+  charging_status_box_.get(charging_status_msg);
+
+  if (!charging_status_msg) {
+    throw opennav_docking_core::FailedToCharge("Did not receive charging_status_msg message");
+  }
+
+  return charging_status_msg->charging;
+}
+
+bool PantherChargingDock::disableCharging()
+{
+  if (!panther_utils::common_utilities::IsPantherVersionAtLeast(panther_version_, 1.2f)) {
+    return true;
+  }
+
+  try {
+    setChargerState(false);
+  } catch (const std::runtime_error & e) {
+    RCLCPP_ERROR_STREAM(logger_, "An exception occurred while disabling charging: " << e.what());
+    return false;
+  }
+
+  return true;
+}
+
+bool PantherChargingDock::hasStoppedCharging()
+{
+  if (!panther_utils::common_utilities::IsPantherVersionAtLeast(panther_version_, 1.2f)) {
+    return !isCharging();
+  }
+
+  std::shared_ptr<panther_msgs::msg::ChargingStatus> charging_status_msg;
+  charging_status_box_.get(charging_status_msg);
+  if (!charging_status_msg) {
+    throw opennav_docking_core::FailedToCharge("Did not receive charging_status_msg message");
+  }
+
+  if (charging_status_msg->charging) {
+    throw opennav_docking_core::FailedToCharge("Charging status is still true");
```

Why we throw instead of return false?